File: 088_Validate Binary Search Tree.py
"""
Validate Binary Search Tree
Check if a binary tree is a valid BST using min/max bounds.
"""
from typing import Optional

class TreeNode:
    def __init__(self,val=0,left=None,right=None):
        self.val=val; self.left=left; self.right=right

def is_valid_bst(root: Optional[TreeNode], low=float('-inf'), high=float('inf')) -> bool:
    if not root: return True
    if not (low < root.val < high): return False
    return is_valid_bst(root.left, low, root.val) and is_valid_bst(root.right, root.val, high)

if __name__ == '__main__':
    root = TreeNode(2)
    root.left = TreeNode(1)
    root.right = TreeNode(3)
    print('Is valid BST:', is_valid_bst(root))
    bad = TreeNode(5)
    bad.left = TreeNode(1)
    bad.right = TreeNode(4)
    bad.right.left = TreeNode(3)
    bad.right.right = TreeNode(6)
    print('Is valid BST:', is_valid_bst(bad))

# Time Complexity: O(n)
# Space Complexity: O(h)

File: 088_Validate Binary Search Tree.cpp
// Validate Binary Search Tree
// Determine if a binary tree is a valid BST.
#include <bits/stdc++.h>
using namespace std;
struct TreeNode{long long val; TreeNode* left; TreeNode* right; TreeNode(long long x): val(x), left(nullptr), right(nullptr) {}};

bool isValidBST(TreeNode* root, long long low=LLONG_MIN, long long high=LLONG_MAX){
    if(!root) return true;
    if(root->val<=low || root->val>=high) return false;
    return isValidBST(root->left, low, root->val) && isValidBST(root->right, root->val, high);
}

int main(){
    TreeNode* root = new TreeNode(2);
    root->left = new TreeNode(1);
    root->right = new TreeNode(3);
    cout<<boolalpha<<"Is valid BST: "<<isValidBST(root)<<"\n";
    TreeNode* bad = new TreeNode(5);
    bad->left = new TreeNode(1);
    bad->right = new TreeNode(4);
    bad->right->left = new TreeNode(3);
    bad->right->right = new TreeNode(6);
    cout<<"Is valid BST: "<<isValidBST(bad)<<"\n"; // false
    return 0;
}

// Time Complexity: O(n)
// Space Complexity: O(h)
